pisa/stages/flux/pi_mceq_barr.py

```diff
@@ -105,267 +186,339 @@ def __init__(self,
-                                  container['barr_b+'].get(WHERE), container['barr_b-'].get(WHERE), barr_b,
-                                  container['barr_c+'].get(WHERE), container['barr_c-'].get(WHERE), barr_c,
-                                  container['barr_d+'].get(WHERE), container['barr_d-'].get(WHERE), barr_d,
-                                  container['barr_e+'].get(WHERE), container['barr_e-'].get(WHERE), barr_e,
-                                  container['barr_f+'].get(WHERE), container['barr_f-'].get(WHERE), barr_f,
-                                  container['barr_g+'].get(WHERE), container['barr_g-'].get(WHERE), barr_g,
-                                  container['barr_h+'].get(WHERE), container['barr_h-'].get(WHERE), barr_h,
-                                  container['barr_i+'].get(WHERE), container['barr_i-'].get(WHERE), barr_i,
-                                  container['barr_w+'].get(WHERE), container['barr_w-'].get(WHERE), barr_w,
-                                  container['barr_y+'].get(WHERE), container['barr_y-'].get(WHERE), barr_y,
-                                  container['barr_z+'].get(WHERE), container['barr_z-'].get(WHERE), barr_z,
-                                  out=container['sys_flux'].get(WHERE),
-                                 )
-            container['sys_flux'].mark_changed(WHERE)
+            #
+            # Apply the systematics to the flux
+            #
 
+            apply_sys_vectorized(
+                container["true_energy"].get(WHERE),
+                container["true_coszen"].get(WHERE),
+                delta_index,
+                energy_pivot,
+                container['nu_flux_nominal'].get(WHERE),
+                container['gradients'].get(WHERE),
+                self.gradient_params,
+                out=container['nu_flux'].get(WHERE),
+            )
 
-@myjit
-def delta(param, pos_f, pos_d, neg_f, neg_d):
-    ''' return fractional delta given a barr parameter and
-    the postitive and negative fluxes and derivateives
-    '''
-    return param * ((pos_d / pos_f) + (neg_d / neg_f))
-
-@myjit
-def mod_factor(idx,
-               barr_a_pos, barr_a_neg, barr_a,
-               barr_b_pos, barr_b_neg, barr_b,
-               barr_c_pos, barr_c_neg, barr_c,
-               barr_d_pos, barr_d_neg, barr_d,
-               barr_e_pos, barr_e_neg, barr_e,
-               barr_f_pos, barr_f_neg, barr_f,
-               barr_g_pos, barr_g_neg, barr_g,
-               barr_h_pos, barr_h_neg, barr_h,
-               barr_i_pos, barr_i_neg, barr_i,
-               barr_w_pos, barr_w_neg, barr_w,
-               barr_y_pos, barr_y_neg, barr_y,
-               barr_z_pos, barr_z_neg, barr_z,
-               ):
-    '''
-    calculate the modification factor for the flux
+            # Check for negative results from spline
+            negative_mask = container['nu_flux'].get(WHERE) < 0
```

Comment:
these are CPU operations as defined here. They should come after `container['nu_flux'].mark_changed(WHERE)` and then be executed on the host like: `negative_mask = container['nu_flux'].get('host') < 0` etc, followed by `.mark_changed('host')`

Reply:
done